File: algorithms/problem-solving/product-sum/product-sum.go
package main

import (
	"fmt"
)

type SpecialArray []interface{}

func ProductSum(xs SpecialArray) int {
	return fn(xs, 1)
}

func fn(array []interface{}, m int) int {

	sum := 0
	for _, x := range array {
		if cast, ok := x.(SpecialArray); ok {
			sum += fn(cast, m+1)
		} else if cast, ok := x.(int); ok {
			sum += cast
		}
	}

	return sum * m
}

func main() {

	xs := SpecialArray{5, 2, SpecialArray{7, -1}, 3, SpecialArray{6, SpecialArray{-13, 8}, 4}}


	res := ProductSum(xs)
	fmt.Println(res)
}

File: algorithms/problem-solving/product-sum/product-sum.ts
type SpecialArray = Array<number | SpecialArray>

export function productSum(array: SpecialArray, multiPlyer = 1): number {
  let sum = 0

  for (const n of array) {
    if (Array.isArray(n)) {
      sum += productSum(n, multiPlyer + 1)
      console.log(sum)
    } else {
      sum += n
    }
  }
  return sum * multiPlyer
}

const xs = [5, 2, [7, -1], 3, [6, [-13, 8], 4]]

console.log(productSum(xs))
